Working Around Iterators as a Generic Type Parameter

chiguire opened the pull request:
This is a snippet for the Haxe Cookbook related to the use of abstracts for having code that can compile in any target, given Haxe's limitation on generic type parameters. Generic type parameters only accept classes or enum instances, not structures, which excludes the possibility of using Iterator out of the box. However, by wrapping Iterator on an abstract, it's possible to work around this limitation.

markknol commented on assets/content/cookbook/Abstract types/working-around-iterator-limitations-as-generic-type-parameter.md:
```diff
@@ -0,0 +1,37 @@
+[tags]: / "abstract-type,type-params,iterator"
+
+# Working Around Iterators as a Generic Type Parameter
```
"Working around" sounds like an hack, lets leave those two words out?  
@Simn ?

markknol replied:
I also think the article belongs to "iterators", not "abstract types", not sure though

chiguire replied:
I tagged it as an abstract type article because I think Haxe's concept of it is less well known than iterators.

markknol commented on assets/content/cookbook/Abstract types/working-around-iterator-limitations-as-generic-type-parameter.md:
```diff
@@ -0,0 +1,37 @@
+
+# Working Around Iterators as a Generic Type Parameter
+
+[Iterators](https://haxe.org/manual/lf-iterators.html) are a helpful Haxe structure. They allow you to build a for loop, but sometimes you want to do something a bit more complex, like iterating over a list of items over time, which excludes the possibility of using a for loop. Ideally you would build a generator function, but Haxe does not have these out of the box. There are libraries that allo ...
```
Also maybe link to a specific example of where/how it is used in moon-core?

markknol commented on assets/content/cookbook/Abstract types/working-around-iterator-limitations-as-generic-type-parameter.md:
```diff
@@ -0,0 +1,37 @@
+
+[Iterators](https://haxe.org/manual/lf-iterators.html) are a helpful Haxe structure. They allow you to build a for loop, but sometimes you want to do something a bit more complex, like iterating over a list of items over time, which excludes the possibility of using a for loop. Ideally you would build a generator function, but Haxe does not have these out of the box. There are libraries that allo ...
+
+However, Iterator is a structure, which is a limitation when you want to use it as the parameter for a generic type. For example, haxe.ds.[GenericStack](http://api.haxe.org/haxe/ds/GenericStack.html)<Iterator<Int>> is an interesting piece of code. It will compile for an html5 target, given Javascript's extreme type flexibility, but it will not compile for other targets, including Flash. The compi ...
```
Why is Flash mentioned here specifically?

chiguire replied:
Well, I'm working with javascript, and I found the error first when trying to compile to cpp. I thought that flash being similar to js could be more flexible and accept the code, but it didn't. I guess it's just the way I managed to find the error and figure out a solution.

markknol commented on assets/content/cookbook/Abstract types/working-around-iterator-limitations-as-generic-type-parameter.md:
````diff
@@ -0,0 +1,37 @@
+
+```haxe
+intStack = new GenericStack<AbstractIterator<Int>>();
+intStack.add(new AbstractIterator(gimme_an_array_of_ints().iterator()));
````
I think "gimme_an_array_of_ints" needs to be explained. Ideally there is a copy/pastable version of this article  (you could use a try-haxe.org snippet).

markknol commented on assets/content/cookbook/Abstract types/working-around-iterator-limitations-as-generic-type-parameter.md:
````diff
@@ -0,0 +1,37 @@
+intStack.add(new AbstractIterator(gimme_an_array_of_ints().iterator()));
+var intStackHasNext = intStack.first().hasNext();
+var intStackNext = intStack.first().next();
+```
````
Don't we also need a `for`loop in the example too?

chiguire commented:
Thanks @markknol, your commentary gives the appropriate accuracy to some things I wasn't clear about, I'll make some changes.

chiguire commented:
Hi @markknol , I rewrote the recipe taking into account your observations. The code in the article works in Try Haxe.

I removed the bit that mentions the generator functions as I feel that could be its own recipe.

markknol commented:
Just asking a stupid question, but why or in which case would you use such structure if the same can be accomplished with just a `for` loop? 
```haxe
var turtles = [A_TURTLE("Leonardo"), MANY_TURTLES([A_TURTLE("Raphael1"), A_TURTLE("Raphael2")]), A_TURTLE("Michelangelo"), A_TURTLE("Donatello")];

function iterate(turtles:Array<Turtles>) {
  for(turtle in turtles) {
    switch (turtle) {
      case A_TURTLE(name): trace("Training the next turtle: " + name);
      case MANY_TURTLES(turtles): iterate(turtles);
    }
  }
}
iterate(turtles);
```
This works on all targets without need of abstracts, type parameters etc.

chiguire commented:
I understand your question, not stupid at all. If you need to iterate over all elements in an array just use a for loop. You use iterators when you need to iterate over these elements in time; when you need to call the function several times because you want to show the intermediate states. Take, for example, that the update loop calls a delay function so there is time to show the progress of the turtle training to the user:

    do {
      exitRequested = updateLoop(turtleStack);
      //showScreenStatusAndDelayFor1000ms();
    } while (!exitRequested);

The update loop then cannot iterate over all the elements at once. The first time updateLoop() is called it will train the first turtle, the second time will train the second turtle, and so on.

I used this technique not to train turtles, but to implement a sequencing language with the enum, in such way that I can have an array of actions that happen in time. The game in a given update tick it will read the next action to do, execute it and then read the next one in the tick after. See [this](https://github.com/chiguire/vzla2017/blob/8daaafb3349e235c68ef25a9fe707d772928085d/source/PlayState.hx#L185) for a more complex implementation of the article. It uses [this enum](https://github.com/chiguire/vzla2017/blob/8daaafb3349e235c68ef25a9fe707d772928085d/source/play/enums/GameActionE.hx), and an [example of such list of actions](https://github.com/chiguire/vzla2017/blob/8daaafb3349e235c68ef25a9fe707d772928085d/source/scenario/Fajardo.hx#L257).

markknol commented:
Sorry for long response. I think the article is good as is. Ill merge it and do some small fixes on it (add you as author for example 👍 )

Thanks for your contribution!

chiguire commented:
Thanks for letting me know, Mark! :)

On 21 November 2017 at 08:10, Mark Knol <notifications@github.com> wrote:

> Sorry for long response. I think the article is good as is. Ill merge it
> and do some small fixes on it (add you as author for example 👍 )
>
> Thanks for your contribution!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/HaxeFoundation/code-cookbook/pull/84#issuecomment-345948236>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACEt9JvYnydsQWdsLiy4nq8cCEPXHGIks5s4oWLgaJpZM4Nf05u>
> .
>
